AWS auth v4

mrwsr opened the pull request:
@exarkun Here it is!  I believe it is the best evidence in favor of TLS client certificates the world has ever seen.

The new module, `_aws_v4_auth`, should have 100% test coverage.

In addition to unit tests, its coverage includes a subset of the cases in [AWS's Auth V4 test suite](http://docs.aws.amazon.com/general/latest/gr/signature-v4-test-suite.html).  The test suite is rather large so these tests are skipped unless you set the `AWS4_TEST_SUITE_PATH` environment variable to the unzipped archive.  Is there a better way to communicate such a path to trial?

I've also modified the Travis configuration to pull down that archive.  We probably don't want to do that because connection issues will make the tests flaky.  I'm open to ideas for improvement.

It may be the case that these tests are not very helpful, too; where the canonical request claims it wants a full URI, it *actually* wants just the path and query string.

I'll fix the merge conflicts later.  I didn't want to wait any longer to push this PR.

exarkun commented on .travis.yml:
```diff
@@ -22,10 +22,13 @@ addons:
 
 script:
-  - "trial txaws"
+  - "wget 'https://docs.aws.amazon.com/general/latest/gr/samples/aws4_testsuite.zip'"
```
I have a bias towards curl.  Not sure if it is rational or not.

Perhaps more productively, if you look above a little bit you can see some caching that gets set up for pip.  I think we can have travis cache this download (or even the unzipped version of it).  If we also have just a little bit of retry logic for failed/interrupted downloads, I think it might actually be fairly reasonable to download it on each travis build.

But on the other hand, this isn't actually a _very_ large archive (a txAWS git checkout is already >7MiB) .  It might also be completely acceptable to just add it to the txAWS repository.  The license is compatible enough so I don't think there are any problems on that front either.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+"""AWS authorization, version 4"""
```
Can you add this line to the top:

```
# Licenced under the txaws licence available at /LICENSE in the txaws source.
```

and uncuddle the docstring quotes?

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+import hmac
+
+import urlparse
+import urllib
```
I'd probably forgo the blank lines between stdlib imports and put the attr import (third part) after them.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+    Produce a SHA-256 HMAC for a message.
+
+    @param key: The secret key to use.
+    @type key: L{str}
```
For maximum clarity, new code should probably talk about either `bytes` or `unicode`.

mrwsr replied:
OK!  I was shooting for consistency but clarity is way more important.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+    @return: The canonicalized headers.
+    @rtype: L{str}
+    """
+    pairs = [(name.lower(), str(headers[name]).strip())
```
a generator expression would also work here, and avoids leaking `name` into the function scope.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+             for name in headers_to_sign
+             if name in headers]
+    sorted_pairs = sorted('{}:{}'.format(name, value)
+                          for name, value in sorted(pairs))
```
`format` is an annoying little nuisance.  The choice to make it a unicode-only method in Python 3.x doesn't _directly_ impact txAWS since txAWS does not support Python 3... Still, using `format` on a byte string grates.  I tend to either stick with `%`-style interpolation or jump through hoops like `u"{}:{}".format(...).encode("utf-8")` (or whatever other encoding is appropriate).

The code is probably fine exactly as it is (in fact, I'm sure it works exactly as desired on Python 2.7 - probably the only version of txAWS cares about).  And someone will have to do a full, careful pass over txAWS if they ever want to port it to Python 3.x _anyway_...

So maybe this is a pointless comment.

mrwsr replied:
This stuck in my craw while writing the code.  You're certainly right that porting txAWS will require serious effort, so it'd be nice if new code didn't make it any harder to port.  At the same time, I don't think I could write Python 3 compatible code without testing it under Python 3, so I didn't want to write something that *appeared* to work but didn't actually.

I'm just now remembering that Python 3.5 and above has byte string `%` formatting, though.  I'll change this (and the other code in this module) to use byte strings, and if/when txAWS does support Python 3, we can just say the minimum supported version is 3.5.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+                    if header in headers)
+
+
+@attr.s
```
Looks like this could be `(frozen=True)` if we wanted.

mrwsr replied:
That implies a slight performance penalty which I was reluctant to pay up front.  But I really like the guarantee that immutability provides.  I'm OK adding it in.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+
+        @param signing_key: The signing key.  Not just your secret
+            key!  See L{getSignatureKey}
+        @type: L{str}
```
Good to have this warning.  It's so easy to mix up the different opaque byte strings involved.  Thanks!

One step further we could go (but maybe shouldn't do _now_) is to give different types to the different types of strings involved.

mrwsr replied:
YES!  The fact that everything becomes a string/bytes is *terrible.*  Now that Twisted's named constants code lives in its own package (thanks, @hawkowl!), we can look forward to converting some of this stuff into constants, and some of this stuff into types...

But not in this PR, which is already gigantic.

exarkun commented on txaws/_auth_v4.py:
```diff
@@ -0,0 +1,433 @@
+    Construct an AWS version 4 authorization value for use in an
+    C{Authorization} header.
+
+    @param region: The AWS region name (e.g., C{us-east-1}).
```
`C{"us-east-1"}` (with the quotes) probably

exarkun commented on txaws/tests/test_auth_v4.py:
```diff
@@ -0,0 +1,723 @@
+                       credential_scope=scope)
+
+
+class AWS4FunctionTest(unittest.SynchronousTestCase):
```
txAWS naming convention seems to be "FooTestCase".

exarkun commented on txaws/tests/test_auth_v4.py:
```diff
@@ -0,0 +1,723 @@
+        self.assertEqual(canonical, textwrap.dedent("""\
+        other-signed-header:other-signed-value
+        signed-header:signed-value
+        """))
```
Another idiom that avoids relying on `textwrap.dedent` to actually do what it's supposed to do is:

```
self.assertEqual(
    canonical,
    "other-signed-header:other-signed-value\n"
    "signed-header:signed-value\n"
)
```

Either is fine, just mentioning it.

exarkun commented on txaws/tests/test_auth_v4.py:
```diff
@@ -0,0 +1,723 @@
+                         "date stamp/region/service/aws4_request")
+
+
+class _CredentialTests(unittest.SynchronousTestCase):
```
I typically drop the leading `_` when naming the tests for a private API.  It's ambiguous about whether it means the class itself is private.  It's typically clear what the unit under test is even without the `_` and the class docstring spells it out if not.

exarkun commented on txaws/tests/test_auth_v4.py:
```diff
@@ -0,0 +1,723 @@
+        self._test_case(self.path.child("get-vanilla-query-order-key-case"))
+
+    def test_get_vanilla_utf8_query(self):
+        self._test_case(self.path.child("get-vanilla-utf8-query"))
```
If you _want_, you could probably write a `test_suite` function that collects all of the cases (using `FilePath.children()`) and generates a `TestSuite` that includes tests for all of them.

But the code is also fine as-written.  And I'd really like to have the authv4 support merged.  So this might be a good follow-up issue. :)

exarkun replied:
Also, I wonder why you didn't choose to exercise all of the tests from the AWS suite.  It looks like there's about 25 of them - why only run this subset of 9?

mrwsr replied:
> If you want, you could probably write a test_suite function that collects all of the cases (using FilePath.children()) and generates a TestSuite that includes tests for all of them.

I erred on the side of obvious but repetitious.  There are also annoying inconsistencies in the directory layout (e.g., `normalize-path` and `post-sts-token` both contain sub-directories with more specific tests). Now that I'm reviewing the tests, though, I'm worried that it's too easy to copy an existing method and rename it, but then forget to change the path name.  I'll give myself half an hour to implement a `test_suite` function and we'll see what ends up in a future commit :)

> Also, I wonder why you didn't choose to exercise all of the tests from the AWS suite. It looks like there's about 25 of them - why only run this subset of 9?

Good question!  Not all the tests pass because of irrelevant parsing issues.  `get-header-key-duplicate`, for example, fails because my crummy "HTTP" parser doesn't handle duplicate headers.  Now, you might think this is the perfect time to use `twisted.web.http.HTTPChannel`, but the canned HTTP requests are not actually HTTP requests.  They don't contain `Content-Length` headers even when a body exists, so an `HTTPChannel` instance can never extract it.  ~I didn't want to make the diff even bigger by either reaching into a private `HTTPChannel` API to grab the suspended body out of some buffer, or by writing a more robust parser that would require its own tests.~ 

On second thought, I can use `HTTPChannel` to parse the headers and then manually extract the body by looking for a blank line.  Will do so and report back!

It gets worse!  `urllib.urlencode` does not implement [RFC 3986's required behavior for unreserved characters](https://tools.ietf.org/html/rfc3986#section-2.3), quoting, say, `~` to `%7E` when it should not.  This causes `get-unreserved` to fail.  I don't see a way around this other than writing a quoting implementation that does the right thing, making the PR *even bigger*.

The tests I did choose to run seemed most relevant to the S3 client.   But now that I've got a better strategy for parsing the requests, I'll try to get more running in within a second half-hour time limit.  Otherwise I'd like to put together a later PR that does so.

markrwilliams commented on txaws/s3/client.py:
```diff
@@ -575,11 +577,13 @@ def __init__(self, bucket=None, object_name=None, data="",
         self.endpoint.set_method(self.action)
 
+    def _utcnow(self):
+        return datetime.datetime.utcnow()
```
+1 for dependency injection!

mrwsr commented on txaws/s3/client.py:
```diff
@@ -596,89 +600,54 @@ def get_headers(self):
-        headers = {"Content-Length": content_length,
-                   "Date": self.date}
+            content_length = str(len(self.data))
+        headers = {"Content-Length": content_length}
```
Nope!  The only headers you have to sign when using V4 auth with S3 are `host` and whatever the date header is (`x-amz-date` here).  I'll definitely get rid of this!

mrwsr commented on txaws/tests/test_aws4_testsuite.py:
```diff
@@ -0,0 +1,308 @@
+_UNRESERVED_SKIP = "urllib.quote quotes unreserved characters"
+_NOT_YET_SUPPORTED_SKIP = "Not yet supported."
+
+_SKIPS = {
```
@exarkun These 12 tests are skipped while the remaining 19 pass.  Later pull requests can whittle this down until it's entirely gone.

mrwsr commented on txaws/tests/test_aws4_testsuite.py:
```diff
@@ -0,0 +1,308 @@
+
+    if not methods:
+        empty_tests = test_class()
+        empty_tests.skip = "Missing AWS test suite directory: {}".format(
```
I can convert this to `%` formatting!

mrwsr commented on txaws/s3/client.py:
```diff
@@ -575,11 +577,13 @@ def __init__(self, bucket=None, object_name=None, data="",
         self.content_type = content_type
         self.metadata = metadata
         self.amz_headers = amz_headers
-        self.date = datetimeToString()
```
I [copied](https://github.com/twisted/txaws/pull/16/files#diff-21760a2cef7b719cd1abf436bb92fcfbR588) an [existing deprecation pattern](https://github.com/twisted/txaws/blob/26bf91da47007ed91ec3598c40a013067f685447/txaws/client/base.py#L195-L209). One thing to note about my approach is that `date` is now useless *and* inaccurate.  This means it's not *really* a deprecation since it's already wrong to call it!  However, it's hard to set the time in the initializer when it's passed, in after instantiation, to the `submit` method, and I do think that's the correct design.  I'm hoping that developers who find their code broken turn on warnings, and if not, that the clear indication of deprecation leads them on the right path.

mrwsr commented:
@exarkun I think I addressed most, if not all, of your (excellent) review comments.  I also fixed a bug that broke at least `put_object`.  Please review again and let me know what else needs to be done.

Thanks!

mrwsr commented:
Also, the S3 client will no longer authenticate correctly when given a body producer.  I'm not sure if that ever worked, though.

exarkun commented:
If there wasn't a test for it, it's hard to argue it ever worked. :)  Offhand I'm not sure which S3 operations require a request body.  I'll probably double check that LeastAuthority is depending on that behavior - most likely not until Monday, though (when I'll re-review and hopefully merge!).  Thanks again!

mrwsr commented:
@exarkun Body producers *can* work so long as they define a `length`.  They *won't* work right now because there's no way to calculate their SHA 256 hash for inclusion in the `x-amz-content-sha256` header. 

It turns out that txAWS currently uses a body producer to upload immediate data, wrapping it in a [`FileBodyProducer`](https://github.com/twisted/txaws/blob/26bf91da47007ed91ec3598c40a013067f685447/txaws/client/base.py#L223-L224).  (txAWS ships with its own `FileBodyProducer` presumably because it predates an era where one shipped with Twisted.  We're in such an era now, though, so we should definitely delete the one in txAWS!). 

So it's certainly the case that some body producers work.  There are two reasons why:

1. the data was available during the `get_headers` call, so its SHA-256 hash could be computed and stashed in the `x-amz-content-sha256` header.  This will not be true for any user-provided body producer.
2. `FileBodyProducers` define a `length` attribute, which means `t.w._newclient.Request.writeTo` (called by `Agent`) [won't use chunked encoding](https://github.com/twisted/twisted/blob/d1b91b92516aa23b52de68d592c1edd30f757072/src/twisted/web/_newclient.py#L811-L814).

Chunked encoding with auth v4 is a nightmare because [each chunk has to be signed via extensions](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html#sigv4-chunked-body-definition), a [feature](https://tools.ietf.org/html/rfc7230#section-4.1.1) that basically nothing implements.  [Twisted certainly doesn't](https://github.com/twisted/twisted/blob/d1b91b92516aa23b52de68d592c1edd30f757072/src/twisted/web/_newclient.py#L1234-L1244) and it would take significant thought to work out how to even communicate this downward.

The good news is that AWS makes chunked encoding basically useless because [you have to specify the total length in the `Content-Length` header!](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html) (see note.)  If that's available, then Twisted won't use chunked encoding, and if it isn't, you can't do chunked encoding.  A very useful Catch-22.

We do need to work out a way for users to communicate the SHA-256 of `length`-ified body producers.  I hope Least Authority's use of the S3 client means this can wait til a later PR. :)

exarkun commented:
Argh, I hadn't fully appreciated how much of a mess the handling of body producers was here.

Fortunately it looks like LeastAuthority (by way of Tahoe-LAFS) doesn't use body_producer anywhere in its S3 code.

So I think we won't have to worry about those issues for this PR.  Someone (maybe me) should file an issue for cleaning up the awful `get_page` interface and perhaps another for adding some safety features around body_producer (rejecting one without a length outright seems like it might be a good idea, for example).
